controller/cmd/service-mirror/probe_worker.go

```diff
@@ -0,0 +1,140 @@
+package servicemirror
+
+import (
+	"fmt"
+	"math/rand"
+	"net/http"
+	"sync"
+	"time"
+
+	logging "github.com/sirupsen/logrus"
+)
+
+const httpGatewayTimeoutMillis = 50000
+
+// ProbeWorker is responsible for monitoring gateways using a probe specification
+type ProbeWorker struct {
+	*sync.RWMutex
+	probe          *GatewayProbeSpec
+	pairedServices map[string]bool
+	stopCh         chan struct{}
+	metrics        *probeMetrics
+	log            *logging.Entry
+}
+
+// NewProbeWorker creates a new probe worker associated with a particular gateway
+func NewProbeWorker(probe *GatewayProbeSpec, metrics *probeMetrics, probekey string) *ProbeWorker {
+	return &ProbeWorker{
+		RWMutex:        &sync.RWMutex{},
+		probe:          probe,
+		pairedServices: make(map[string]bool),
+		stopCh:         make(chan struct{}),
+		metrics:        metrics,
+		log: logging.WithFields(logging.Fields{
+			"probe-key": probekey,
+		}),
+	}
+}
+
+// NumPairedServices returns the number of paired services for this probe worker
+func (pw *ProbeWorker) NumPairedServices() int {
+	return len(pw.pairedServices)
+}
+
+// PairService increments the number of services that are routed by the gateway
+func (pw *ProbeWorker) PairService(serviceName, serviceNamespace string) {
+	svcKey := fmt.Sprintf("%s-%s", serviceNamespace, serviceName)
+	if _, ok := pw.pairedServices[svcKey]; !ok {
+		pw.pairedServices[svcKey] = true
+		pw.metrics.services.Inc()
```

Comment:
do you think it would be more robust to do something like `pw.metrics.services.SetValue(len(pw.pairedServices))`?